### PR TITLE
layout and form metadata merging

### PR DESCRIPTION
**Pull request by AlmogShaul**

### CODE REVIEW

File: src/app/types/types.ts [hunk 1]
```diff
@@ -1,51 +1,63 @@
...
+export interface Manifest {
+  asset: Asset[];
 }
 
-export interface Asset{
-  group:Group;
-  assetType:string;
-  asset:Asset[];
-  name:string;
+export class FormLayout {
```
Comment (schmuli) [inline src/app/types/types.ts]: The following doesn't include 'collapsible'.

If we consider that layout is being used to extend the manifest JSON to include properties that `sf-form` supports but which are either 1) non-standard JSON Schema, or 2) unknown from the manifest, we could consider the Layout schema to just be those missing properties.

If the layout schema will just use the missing sf-form schema properties, wouldn't that allow:

1. a much simpler merge of schema and layout (it would still require a deep merge)
2. not having to worry about cloning, since merging twice would always result in the same schema
3. if we add some layout property to the schema, for example fieldsets, nothing would need to change

What do you think?

File: src/app/eform/services/transformation.service.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
+import {Injectable} from '@angular/core';
+import { Asset, Dependency, FactType, FormElement, FormLayout, FormSchema, Manifest } from '../../types/types';
+
+
+@Injectable()
+export class TransformationService {
```
Comment (schmuli) [inline src/app/eform/services/transformation.service.ts]: I think we should split this service into two. You already have 200 lines here, and these are 2 distinct features

File: src/app/eform/preview/preview.component.ts [hunk 3]
```diff
@@ -1,43 +1,62 @@
...
-      }
-      catch (er) {
-        this.error = PARSE_ERROR;
-      }
-
-    },error=>{
-      this.error = error;
-    });
-  }
+    filesSelected(data: any) {
```
Comment (schmuli) [inline src/app/eform/preview/preview.component.ts]: I don't think we should keep this behavior.

1. It is not intuitive that selecting two files will work, and especially that if one is called `layout` it will be treated differently
2. It should support a pre-configured layout file (such as on the server)

The simplest thing for now is to have 2 file select inputs, one for manifest and one for layout (optional), and add a Preview button to actually bind into the form.

File: src/app/eform/preview/preview.component.html [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
-      <h4>Select manifest file</h4>
+      <h4>Select manifest and layout files</h4>
       <div class="fileUpload btn btn-info"  >
         <span>Upload</span>
-        <input type="file" (change)="manifestFileSelected($event)" class="upload">
+        <input type="file" (change)="filesSelected($event)" class="upload" multiple>
       </div>
     </div>
+
     <h3>{{error}}</h3>
```
Comment (schmuli) [inline src/app/eform/preview/preview.component.html]: I think a better way to display an error is to have the message in the HTML, and just set an `error` flag so that it is displayed. This is what I showed you today with the Process Selection page.

File: src/app/eform/eform/eform.component.ts [hunk 5]
```diff
@@ -1,27 +1,37 @@
...
 
+  constructor(private transformationService:TransformationService){}
 
-    ngOnChanges(changes: any): void {
-        this.schema = changes.metadata.currentValue;
+    ngOnChanges(): void {
+        this.schema = null;
+        if (this.metadata) {
+            //workaround to activate the CD when a new metadata or layout arrived
+            setTimeout(()=>{
```
Comment (AlmogShaul) [inline src/app/eform/eform/eform.component.ts]: I don't agree, This component saves us rewriting the same code over and over again, every component that will need form, will have to transform its manifest and merge its layout.
about your  last comment, I agree we need to think of caching in farther releases.

File: src/app/types/types.ts (hunk 1 above)
Comment (AlmogShaul) [inline src/app/types/types.ts]: "If the layout schema will just use the missing sf-form schema properties" - this is exactly what I did in the FormLayout.
Can you more clear ?

File: src/app/eform/services/transformation.service.ts (hunk 2 above)
Comment (AlmogShaul) [inline src/app/eform/services/transformation.service.ts]: I don't agree. those 2 features related to data transformation/merging , maybe I will change the service name

File: src/app/eform/preview/preview.component.ts (hunk 3 above)
Comment (AlmogShaul) [inline src/app/eform/preview/preview.component.ts]: ok

File: src/app/eform/preview/preview.component.html (hunk 4 above)
Comment (AlmogShaul) [inline src/app/eform/preview/preview.component.html]: To be honest I am not good with both solutions, I think we need toaster. lets leave this issue for another PR ok?

File: src/app/eform/services/transformation.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            'version': '1.0'
+        }
+    ],
+    'executionModelHash': 'bc776b36d503cea5f1353bdbc67347b0',
+    'release': 'Sample Release',
+    'tag': 'test3',
+    'conclusion': 'Test'
+};
+
+const layout: FormLayout = <FormLayout> {
```
Comment (AlmogShaul) [inline src/app/eform/services/transformation.service.spec.ts]: ok

File: src/app/eform/eform/eform.component.ts (hunk 5 above)
Comment (schmuli) [inline src/app/eform/eform/eform.component.ts]: The only repeated code is the template, which will happen anyway.
Since there is a transformer service, the only duplicate code is setting a component property to the result of a call to the service. Considering that we each have slightly different requirements for how to get a manifest and layout, I don't think that is an issue.

File: src/app/eform/preview/preview.component.html (hunk 4 above)
Comment (schmuli) [inline src/app/eform/preview/preview.component.html]: OK

Comment (schmuli) [inline src/app/eform/preview/preview.component.html]: OK

File: src/app/types/types.ts (hunk 1 above)
Comment (schmuli) [inline src/app/types/types.ts]: I mean that we can use the same property names and structure.

Use `fieldsets` rather than `sections`
Use `required` rather than `requiredProperties`
For dependencies, specify the property and use `visibleIf`option.

For example:

```json
{
    "fieldsets": [
        ...
    ],
    "properties": {
        "prop": { 
            "visibleIf": { "otherProp": [true] }
        }
    }
}
```

Comment (schmuli) [inline src/app/types/types.ts]: This also allows you to define required properties for an inner object definition, not just for the root object.

Comment (AlmogShaul) [inline src/app/types/types.ts]: ah , ok, but I did it to make it more clear to the person who should edit / create the layout file.

File: src/app/eform/services/transformation.service.ts (hunk 2 above)
Comment (schmuli) [inline src/app/eform/services/transformation.service.ts]: I think it might not be all that important, since we will probably move some of the functionality to the server anyway, and we can handle it then

File: src/app/types/types.ts (hunk 1 above)
Comment (AlmogShaul) [inline src/app/types/types.ts]: About the required, I don't think  they implement this inner requires ...

Comment (schmuli) [inline src/app/types/types.ts]: I agree, I am not which is more important. This is just a thought, we can leave it as is for now, maybe look again in a later release

File: src/app/eform/services/transformation.service.spec.ts (hunk 6 above)
Comment (schmuli) [inline src/app/eform/services/transformation.service.spec.ts]: Sorry for the misunderstanding, I only meant the `layout` and `formManifest` constants, since they are small and can be inside the specific `describe`.

The `manifest` can be left where it was, or did you have an issue with it being at the end?

**Comment by schmuli** [comment]
Just look at the comment on the Spec.

I think the Preview component looks great now, even though you have added some more code to it. It is clear what it is doing, and when. Thanks.
